File: linguagem_C/aquecimento_recursivo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int contando (int contador, char* sentenca, int tamanho, int indice){
  if (indice == tamanho){
    return contador;
  }
  else{
    if (sentenca[indice] >= 'A' && sentenca[indice] <= 'Z') { // se o caractere for maiusculo 
      return contando (contador+1, sentenca, tamanho, indice+1);
    }
    else{
      return contando (contador, sentenca, tamanho, indice+1);
    }
  }
}

int main(void) {

  char sentenca[150];
  int quantos_char = 0;
  int quantas_maiusculas = 0;

  while (1) {
    scanf(" %[^\n]", sentenca); // le a linha ate o \n

    quantos_char = strlen(sentenca);

    if (strcmp(sentenca, "FIM") == 0) { // comparando se a minha string eh igual a string FIM
      break; 
    }
    
    printf("%d\n", contando(0, sentenca, quantos_char, 0)); 
  }

  return 0;
}

File: Java/aquecimento_recursivo.java
import java.util.Scanner;

public class aquecimento_recursivo {

    public static int contando(int contador, String sentenca, int tamanho, int indice) {
        if (indice == tamanho) { // se minha string acabo, retorna quantas maiusculas
            return contador;
        } else {
            if (sentenca.charAt(indice) >= 'A' && sentenca.charAt(indice) <= 'Z') { // se o caractere for maiúsculo
                return contando(contador + 1, sentenca, tamanho, indice + 1); // chama denovo subindo um no indice e contador
            } else {
                return contando(contador, sentenca, tamanho, indice + 1); // chama denovo so no indice
            }
        }
    }

    public static void main(String[] args) {
        
        Scanner leitor = new Scanner(System.in);
        String sentenca;
        int quantos_char;

        while (true) {
            sentenca = leitor.nextLine(); // le ate o final da linha

            if (sentenca.equals("FIM")) { // compara se a string é igual a FIM
                break;
            }

            quantos_char = sentenca.length(); // conta quantos caracteres tem a string

            System.out.println(contando(0, sentenca, quantos_char, 0)); // printando o resultado 
        }

        leitor.close(); // fecha meu leitor
    }
}
